book-content/chapters/07-mutability.md

```diff
@@ -0,0 +1,1013 @@
+
+```typescript
+const readOnlyGenres: ReadonlyArray<string> = ["rock", "pop", "unclassifiable"];
+```
+
+Both of these approaches are functionally the same. Hovering over the `readOnlyGenres` variable shows that TypeScript has inferred it as a read-only array:
+
+```typescript
+// hovering over `readOnlyGenres` shows:
+const readOnlyGenres: readonly string[];
+```
+
+Readonly arrays disallow the use of array methods that cause mutations, such as `push` and `pop`:
+
+```ts twoslash
+// @errors: 2339
+const readOnlyGenres: readonly string[] = ["rock", "pop", "unclassifiable"];
+
+// ---cut---
+readOnlyGenres.push("experimental");
+```
+
+However, methods like `map` and `reduce` will still work, as they create a copy of the array and do not mutate the original.
+
+```ts twoslash
+// @errors: 2339
+const readOnlyGenres: readonly string[] = ["rock", "pop", "unclassifiable"];
+
+// ---cut---
+const uppercaseGenres = readOnlyGenres.map((genre) => genre.toUpperCase()); // No error
+
+readOnlyGenres.push("experimental");
+```
+
+Note that, just like the `readonly` for object properties, this doesn't affect the runtime behavior of the array. It's just a way to help catch potential errors.
+
+#### How Read-Only and Mutable Arrays Work Together
+
+To help drive the concept home, let's see how read-only and mutable arrays work together.
+
+Here are two `printGenre` functions that are functionally identical, except `printGenresReadOnly` takes a read-only array of genres as a parameter whereas `printGenresMutable` takes a mutable array:
+
+```typescript
+function printGenresReadOnly(genres: readonly string[]) {
+  // ...
+}
+
+function printGenresMutable(genres: string[]) {
+  // ...
+}
+```
+
+When we create a mutable array of genres, it can be passed as an argument to both of these functions without error:
+
+```typescript
+const mutableGenres = ["rock", "pop", "unclassifiable"];
+
+printGenresReadOnly(mutableGenres);
+printGenresMutable(mutableGenres);
+```
+
+This works because specifying `readonly` on the `printGenresReadOnly` function parameter only guarantees that it won't alter the array's content. Thus, it doesn't matter if we pass a mutable array because it won't be changed.
+
+However, the reverse is not true.
+
+If we declare a read-only array, we can only pass it to `printGenresReadOnly`. Attempting to pass it to `printGenresMutable` will yield an error:
+
+```ts twoslash
+// @errors: 2345
+function printGenresReadOnly(genres: readonly string[]) {
+  // ...
+}
+
+function printGenresMutable(genres: string[]) {
+  // ...
+}
+
+// ---cut---
+const readOnlyGenres: readonly string[] = ["rock", "pop", "unclassifiable"];
+
+printGenresReadOnly(readOnlyGenres);
+printGenresMutable(readOnlyGenres);
+```
+
+This is because we might be mutating the array inside of `printGenresMutable`. If we passed a read-only array.
```

Comment:
That's technically not exactly true. It's not because the function could mutate the array inside. If that were true then this would be illegal too:
```ts
type ReadonlyObj = { readonly foo: string }

function changeObj(obj: { foo: string }) {
  obj.foo = 'bazinga'
}

const obj: ReadonlyObj = { foo: 'bar' }

changeObj(obj) // it would be an error if the above statement would be true
```

The reason why it's illegal to pass a readonly array to a spot expecting a mutable array is purely structural. Just like you mentioned, the readonly ones don't have `push`, `pop` and maybe some other things too. So it's as if you'd try to do this:
```ts
function require2Props(obj: { a: number, b: string }) {}

require2Props({ a: 42 }) // error
```

Reply:
But also... we might actually get a new tsconfig option that will help people to benefit from a stricter set of rules: https://github.com/microsoft/TypeScript/pull/58296